More granular locking on Messenger

Samyoul opened the pull request:
Hopefully addresses issues reported in https://github.com/status-im/status-go/issues/1936.

Samyoul committed:
Added preliminary work for more granual locking on Messenger

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (27)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 8485dc85 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/1/) | 2021-03-29 15:49:00 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/1/consoleText) |
| :x: | 8485dc85 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/1/) | 2021-03-29 15:49:05 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/1/consoleText) |
| :x: | 8485dc85 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/1/) | 2021-03-29 15:49:37 | ~2 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/1/consoleText) |
| | | | | | | |
| :x: | 31a15f80 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/2/) | 2021-03-30 12:38:30 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/2/consoleText) |
| :x: | 31a15f80 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/2/) | 2021-03-30 12:38:55 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/2/consoleText) |
| :x: | 31a15f80 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/2/) | 2021-03-30 12:39:30 | ~2 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/2/consoleText) |
| | | | | | | |
| :heavy_check_mark: | 9c5eb58b | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/3/) | 2021-03-30 15:59:13 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210330-155723-9c5eb5.zip) |
| :heavy_check_mark: | 9c5eb58b | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/3/) | 2021-03-30 16:02:15 | ~4 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210330-155723-9c5eb5.zip) |
| :heavy_check_mark: | 9c5eb58b | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/3/) | 2021-03-30 16:02:46 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210330-155723-9c5eb5.aar) |
| | | | | | | |
| :heavy_check_mark: | 0b133047 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/4/) | 2021-04-08 11:45:53 | ~2 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210408-114337-0b1330.zip) |
| :heavy_check_mark: | 0b133047 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/4/) | 2021-04-08 11:48:11 | ~4 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210408-114337-0b1330.zip) |
| :heavy_check_mark: | 0b133047 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/4/) | 2021-04-08 11:49:19 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210408-114334-0b1330.aar) |
| | | | | | | |
| :heavy_check_mark: | 9faa8291 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/5/) | 2021-04-08 14:53:43 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210408-145035-9faa82.zip) |
| :heavy_check_mark: | 9faa8291 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/5/) | 2021-04-08 14:54:03 | ~3 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210408-145038-9faa82.zip) |
| :heavy_check_mark: | 9faa8291 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/5/) | 2021-04-08 14:55:35 | ~4 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210408-145035-9faa82.aar) |
| | | | | | | |
| :heavy_check_mark: | 4d0e0746 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/6/) | 2021-04-08 15:58:04 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210408-155616-4d0e07.zip) |
| :heavy_check_mark: | 4d0e0746 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/6/) | 2021-04-08 15:58:15 | ~1 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210408-155616-4d0e07.zip) |
| :heavy_check_mark: | 4d0e0746 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/6/) | 2021-04-08 16:01:12 | ~4 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210408-155616-4d0e07.aar) |
| | | | | | | |
| :heavy_check_mark: | 370fca62 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/7/) | 2021-04-08 16:03:02 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210408-160133-370fca.zip) |
| :heavy_check_mark: | 370fca62 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/7/) | 2021-04-08 16:03:05 | ~1 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210408-160133-370fca.zip) |
| :heavy_check_mark: | 370fca62 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/7/) | 2021-04-08 16:06:54 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210408-160131-370fca.aar) |
| | | | | | | |
| :heavy_check_mark: | ed6f3654 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/8/) | 2021-04-09 09:14:41 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210409-091305-ed6f36.zip) |
| :heavy_check_mark: | ed6f3654 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/8/) | 2021-04-09 09:14:58 | ~1 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210409-091305-ed6f36.zip) |
| :heavy_check_mark: | ed6f3654 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/8/) | 2021-04-09 09:17:28 | ~4 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210409-091305-ed6f36.aar) |
| | | | | | | |
| :heavy_check_mark: | 5cdad4fa | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/9/) | 2021-04-09 12:57:54 | ~2 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210409-125553-5cdad4.zip) |
| :heavy_check_mark: | 5cdad4fa | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/9/) | 2021-04-09 12:57:57 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210409-125553-5cdad4.zip) |
| :heavy_check_mark: | 5cdad4fa | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/9/) | 2021-04-09 13:01:09 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210409-125550-5cdad4.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 7a76b30f | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/10/) | 2021-04-09 21:31:19 | ~30 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/10/consoleText) |
| :x: | 7a76b30f | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/10/) | 2021-04-09 21:32:03 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/10/consoleText) |
| :x: | 7a76b30f | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/10/) | 2021-04-09 21:32:04 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/10/consoleText) |
| | | | | | | |
| :heavy_check_mark: | d7358915 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2181/11/) | 2021-04-09 21:38:35 | ~1 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-210409-213703-d73589.zip) |
| :heavy_check_mark: | d7358915 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2181/11/) | 2021-04-09 21:38:45 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-210409-213703-d73589.zip) |
| :heavy_check_mark: | d7358915 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2181/11/) | 2021-04-09 21:41:46 | ~4 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-210409-213703-d73589.aar) |

Samyoul committed:
Added mutex update to messenger_chats and 'improved' Locker

We'll see, I think that preventing a global lock from locking before a granular lock has been issued will be helpful.

Samyoul committed:
Replaced mutex usage with locker

Samyoul committed:
Switch of all chat maps to chatMap lock type

Samyoul committed:
Changed contact maps to contactMap locks

Samyoul committed:
Updated system message translation map to a sync map

Samyoul committed:
Implemented installationMap and stringBoolMap

Samyoul committed:
Fix failing tests

Samyoul committed:
Removed customer locker 😞 and most usages of locking in Messenger

Samyoul committed:
Little tweaks

Samyoul committed:
go fmt

Samyoul committed:
Revert MessengerResponse.removedChats sync map change

I don't think it is needed, I think that a MessengerResponse will accessed by only a single routine so current map protection isn't needed

Samyoul committed:
Applying sync.Map.Load to a sneaky new AllContacts call

cammellos commented:
Closing as merged in a different PR, thanks @Samyoul

Samyoul commented:
Merged in with https://github.com/status-im/status-go/pull/2199
